proposals/p2760.md

```diff
@@ -0,0 +1,794 @@
+[conditional conformance](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#conditional-conformance).
+We can work around this restriction by using an non-extended interface
+implementation and aliases to the members of the interface.
+
+What was previously written:
+
+```
+interface Printable {
+  fn Print[self: Self]();
+}
+
+class Vector(T:! type) {
+  // ...
+
+  impl forall [U:! Printable] Vector(U) as Printable {
+    fn Print[self: Self]();
+  }
+}
+```
+
+is now written:
+
+```
+interface Printable {
+  fn Print[self: Self]();
+}
+
+class Vector(T:! type) {
+  // ...
+  alias Print = Printable.Print;
+}
+
+impl forall [U:! Printable] Vector(U) as Printable {
+  fn Print[self: Self]();
+}
+```
+
+The way this works is that `Vector.Print` is equivalent to
+`Vector.(Printable.Print)`, which may or may not be defined. The name
+`Vector.Print` can no longer be conditional, and the meaning of that name is
+fixed. However, the implementation of `Printable` for `Vector(T)` may not exist
+for some types `T`.
+
+### Adapters
+
+What was previously written:
+
+```
+class C;
+// Forward declarations of adapters.
+adapter A for C;
+adapter E extends C;
+
+// Definitions of an adapter.
+adapter A for C {
+  ...
+}
+// Definition of an extending adapter.
+adapter E extends C {
+  ...
+}
+```
+
+is now written:
+
+```
+class C;
+// Forward declarations of adapters.
+class A;
+class E;
+
+// Definitions of an adapter.
+class A {
+  adapt C;
+  ...
+}
+// Definition of an extending adapter.
+class E {
+  extend adapt C;
+  ...
+}
+```
+
+Note:
+
+-   Adapters are now a special case of classes, not a distinct top-level
+    declaration.
+-   Classes with `adapt` still must not contain anything that was previously
+    forbidden for adapters: no fields, no base class, no virtual methods, no
+    implementations of virtual methods, and so on.
+-   The `adapt` declaration must appear before
+    [mixin declarations](#future-work-mixins), if any.
+-   The syntax for an `adapt` declaration inside a class body is:
+
+    > [`extend`] `adapt` _type-expression_ `;`
+
+### Interfaces
+
+What was previously written:
+
+```
+interface A { let T:! Type; }
+interface B { let U:! Type; }
+interface C { let V:! Type; }
+interface D { let W:! Type; }
+
+interface I {
+  // Interface is incorporated into `I`.
+  extends A where .T == i32;
+  extends B where .U == i32;
+
+  // No impact on `I`s interface, but an implementation must exist.
+  impl as C where .V == i32;
+  impl as D where .W == i32;
+}
+```
+
+is now written:
+
+```
+interface A { let T:! Type; }
+interface B { let U:! Type; }
+interface C { let V:! Type; }
+interface D { let W:! Type; }
+
+interface I {
+  // `A`'s interface is incorporated into `I`.
+  extend A;
+
+  // Anything that could be in a `where` clause can
+  // be in a `require` declaration:
+  require A.T == i32;
+
+  // Can combine `extend` with `where`.
+  extend B where B.U == i32;
```

Comment:
```suggestion
  extend B where .U == i32;
```
(In `B where ...` we have `.Self: B`, so `.U`, which is `.Self.U`, refers to `.Self.(B.U)`.)